dan_layer/core/src/workers/hotstuff_waiter.rs

```diff
@@ -698,14 +752,30 @@ where
     // We are running a pacemaker for the proposal. But the other committee is still having an election. Let's postpone
     // the pacemaker.
     async fn on_receive_election_in_progress(
-        &self,
+        &mut self,
         _from: TAddr,
         epoch: Epoch,
         shard_id: ShardId,
         payload_id: PayloadId,
     ) -> Result<(), HotStuffError> {
         // TODO: if there is a byzantine replica, he can always send election in progress. We should count the election
         // rounds. Be aware that the election rounds are based on delta time, and this trigger is 2*delta.
+        let election_rounds = self.elections_rounds.entry((shard_id, payload_id)).or_default();
+        *election_rounds += 1;
+
+        // For now, we hardcode the max num of election rounds as 4, before we cancel the transaction
+        if *election_rounds >= 4 {
```

Comment:
A byzantine node may send 4 messages to get the payload rejected. This can currently also happen without shenanigans if `f + 1 >= 4` sent in response to `f + 1`  `MissingProposal`s. We need to check that we sent the MissingProposal in the first place and this is a response to it which is a bit more complex (Req/Resp semantics, keeping sent request ids and checking that the response matches).  

To fail the transaction, we should set the payload result to rejected and notify using `self.publish_event(HotstuffEvent::Failed(...)` as well as abandon the pledges. 

```rust
 let finalize_result = FinalizeResult::reject(
                    payload_id.into_array().into(),
                    RejectReason::ForeignLeaderFailure(...),
                );

                tx.update_payload_result(payload_id, PayloadResult {
                    finalize_result,
                    pledge_hash,
                })?;
```

Reply:
The logic to make this work is more involved than I anticipated. The tricky part is to get the `pledge_hash` to the `PayloadResult`. A possible way is to check local storage contents for each `involved_shard` in the `payload`, as well as the latest `payload_height`. This should be reviewed one more time, I reckon.